connlimit.go

```diff
@@ -182,11 +193,26 @@ func (l *Limiter) SetConfig(c Config) {
 // their own calls if they need to continue limiting the number of concurrent
 // hijacked connections.
 func (l *Limiter) HTTPConnStateFunc() func(net.Conn, http.ConnState) {
+	message := "Your IP is issuing too many concurrent connections, please rate limit your calls\n"
+	tooManyRequestsResponse := []byte(fmt.Sprintf("HTTP/1.1 429 Too Many Requests\r\n"+
+		"Content-Type: text/plain\r\n"+
+		"Content-Length: %d\r\n"+
+		"Connection: close\r\n\r\n%s", len(message), message))
 	return func(conn net.Conn, state http.ConnState) {
 		switch state {
 		case http.StateNew:
 			_, err := l.Accept(conn)
 			if err != nil {
+				if err == ErrPerClientIPLimitReached {
+					go func() {
+						// We don't care about slow players
+						if l.RWDeadlineMaxDelay > 0 {
+							conn.SetWriteDeadline(time.Now().Add(l.RWDeadlineMaxDelay))
+						}
+						conn.Write(tooManyRequestsResponse)
```

Comment:
AFAICT this will race with Writes by the actual HTTP handler and could create a corrupt response (eg 2 HTTP responses written for 1 request that was over the limit). I would not be surprised if behavior varied depending on Go version since internal http.Server implementation details could affect the outcome.

Since there's no way to share state between the ConnState func and the Handler, I don't think there's a way to synchronize this and fix it.

Reply:
Did not see it (I was testing on the simplest - so faster - endpoint: "/", the endpoint returning "Consul Agent" in plain text), but yes, I suspect that's in theory possible